DOM: Test {mouse,}wheel and touch{move,start} events are non-cancelable when there are no non-passive event listeners

zcorpan opened the pull request:
Test that events are not cancelable when there are no non-passive event listeners, and cancelable when there are non-passive event listeners.

Follows https://w3c.github.io/touch-events/#cancelability

> In particular, a user agent may generate only uncancelable touch events when it [observes that there are no non-passive listeners](https://dom.spec.whatwg.org/#observing-event-listeners) for the event.

...and the same for `wheel` events and `mousewheel` events, but that is not yet defined in https://w3c.github.io/uievents/#cancelability-of-wheel-events . The relevant issue is https://github.com/w3c/uievents/issues/282

zcorpan committed:
Test passive/non-passive wheel events on window, document, body, div

zcorpan committed:
Test passive/non-passive mousewheel events on window, document, body, div

zcorpan committed:
Test passive/non-passive touchmove events on window, document, body, div

zcorpan committed:
Test passive/non-passive touchstart events on window, document, body, div

zcorpan committed:
Remove console.log

zcorpan commented:
* [Demo for mousewheel, touchstart, touchmove on window, document, body](
http://software.hixie.ch/utilities/js/live-dom-viewer/saved/10395) (and [same for wheel](http://software.hixie.ch/utilities/js/live-dom-viewer/saved/10396))
  - Chrome: passive
* [Demo for mousewheel, touchstart, touchmove on div, with `event.preventDefault()`](http://software.hixie.ch/utilities/js/live-dom-viewer/saved/10397) (and [same for wheel](http://software.hixie.ch/utilities/js/live-dom-viewer/saved/10400))
  - Chrome: NOT passive
* [Demo for mousewheel, touchstart, touchmove on div, *without* `event.preventDefault()`](http://software.hixie.ch/utilities/js/live-dom-viewer/saved/10399) (and [same for wheel](http://software.hixie.ch/utilities/js/live-dom-viewer/saved/10398))
  - Chrome: First event in a single scroll gesture is NOT passive, the remaining events *are* passive. I don't see this in the [intervention design doc](https://docs.google.com/document/d/1dLxnzyixmR9MeHXE75Uz2nIVA0PaEjN4W3V_-df0sYI/edit) linked from https://github.com/WICG/interventions/issues/64. Is there a newer intervention that explains this? (This is not yet covered in the tests of this PR.)

I'll test with touch-enabled Firefox and Safari and edit this comment with findings.

zcorpan commented:
I used separate demos for mousewheel and wheel because at least in Chrome, when there were `wheel` event listeners, it would only fire `wheel` events, not `mousewheel` events. I suppose it's because pages can listen for both, but expect only one of them. (Not sure if this is specified, `mousewheel` isn't even mentioned in [UI Events](https://w3c.github.io/uievents/).)

zcorpan committed:
Test with and without preventDefault(), check > 1 event

zcorpan committed:
Move tests into their own directory and rename filesœ

zcorpan committed:
Also test the root element

zcorpan committed:
Change pass condition for touchmove-event-listener-on-div-without-preventDefault

zcorpan commented:
I tried testing multiple events, but could not reproduce passive after non-passive with testdriver.js.

If the concept of default passive event listeners is unrelated to whether a previous event was `preventDefault()`ed, I'll remove that from these tests.

zcorpan committed:
Remove complexity for multiple events and preventDefault

zcorpan committed:
Remove .tentative

domenic commented on dom/events/passive-default/resources/default-passive-scrolling.js:
```diff
@@ -0,0 +1,40 @@
+      // We requestAnimationFrame either for 200 frames or until condition is
+      // met.
+      if (frames >= MAX_FRAME)
+        reject("Condition did not become true after 200 frames");
```
```suggestion
        reject(new Error("Condition did not become true after 200 frames"));
```

This is best practice in general and will ensure you get stack traces on failures.

domenic commented on dom/events/passive-default/resources/default-passive-scrolling.js:
```diff
@@ -0,0 +1,40 @@
+      else if (condition())
+        resolve();
+      else
+        requestAnimationFrame(tick.bind(this, frames + 1));
```
```suggestion
        requestAnimationFrame(() => tick(frames + 1));
```

IMO clearer

domenic commented on dom/events/passive-default/resources/default-passive-scrolling.js:
```diff
@@ -0,0 +1,40 @@
+    event.preventDefault();
+  });
+
+  promise_test (async (t) => {
```
```suggestion
  promise_test(async (t) => {
```

domenic commented on dom/events/passive-default/resources/default-passive-scrolling.js:
```diff
@@ -0,0 +1,40 @@
+  promise_test (async (t) => {
+    t.add_cleanup(() => {
+      document.querySelector('.remove-on-cleanup')?.remove();
+    })
```
```suggestion
    });
```

domenic commented on dom/events/passive-default/resources/default-passive-touching.js:
```diff
@@ -0,0 +1,52 @@
+function waitFor(condition) {
```
Might be nice to factor this out into a shared file...

domenic commented on dom/events/passive-default/resources/default-passive-scrolling.js:
```diff
@@ -0,0 +1,40 @@
+    await new test_driver.Actions()
+      .scroll(pos_x, pos_y, delta_x, delta_y).send();
+    await waitFor(()=> { return arrived; });
+    assert_equals(cancelable, expectCancelable);
```
I think it's also worth testing `event.defaultPrevented`. From what I understand, an event might be `cancelable`, but because of the proposal we're testing here, its listeners will be passive, so even after calling `event.preventDefault()`, `event.defaultPrevented` will be false.

Right now it seems like an implementation could pass the tests by never doing anything regarding passive-ness, only cancelable-ness.

zcorpan replied:
You're right, I had misunderstood now passive works and thought it wasn't directly testable, but indeed it is testable with `defaultPrevented` and firing a cancelable event with `dispatchEvent`, like this:

 http://software.hixie.ch/utilities/js/live-dom-viewer/saved/10420

domenic commented on dom/events/passive-default/resources/default-passive-scrolling.js:
```diff
@@ -0,0 +1,40 @@
+    const delta_y = 100;
+    await new test_driver.Actions()
+      .scroll(pos_x, pos_y, delta_x, delta_y).send();
+    await waitFor(()=> { return arrived; });
```
```suggestion
    await waitFor(() => { return arrived; });
```

domenic reviewed (commented):
So I left an inline comment about passive vs. non-cancelable, but I'm wondering if there's a bigger problem there... or maybe I'm just not understanding.

Basically, is there any evidence that passive-ness is involved here at all? It seems like everything could be explained by browsers just firing non-`cancelable` events. Whereas the point of `passive` is to allow certain _listeners_ to have no-op `preventDefault()`, even if the event can be `preventDefault()`ed by other listeners and thus has `cancelable = true`.

EDIT: maybe the evidence is that if you set `passive: false` explicitly, that overrides the default behavior? But in that case `cancelable` should be true, whereas these tests expect false for non-divs...

zcorpan commented:
This tests that the events are non-cancellable as allowed in https://w3c.github.io/touch-events/#cancelability and ... not yet in https://w3c.github.io/uievents/#cancelability-of-wheel-events . The issue is https://github.com/w3c/uievents/issues/282 -- I can submit a PR to fix the spec.

> In particular, a user agent may generate only uncancelable touch events when it [observes that there are no non-passive listeners](https://dom.spec.whatwg.org/#observing-event-listeners) for the event.

I'm not sure why this is only a "may", and not a "must" or at least "should". I'll file an issue.

domenic commented:
Hmm, so that is very weird. So basically, if you do `addEventListener(..., { passive: false })` explicitly, then the UA will allow `cancelable = true`... but if you do `addEventListener(...)` alone, then `cancelable = false`?

I would have expected `cancelable` to not change, and always be true, but passive listeners would do nothing when they call `preventDefault()`. That is what being a passive listener means, after all! The extra complication of flipping `cancelable` depending on what is going on with the listeners seems really unnecessary...

zcorpan commented:
I guess the browser would be lying if it fired events with `cancelable = true` that are not in fact cancelable. You wouldn't know without registering a non-passive listener, of course. I guess there could be a race condition between the browser queueing an event to be fired (non-cancelable) and the page registering a non-passive listener?

zcorpan commented:
@domenic I've opened a new PR with the strategy discussed in https://github.com/web-platform-tests/wpt/pull/34464#discussion_r908277278 which doesn't rely on the related but separate intervention to fire cancelable events when there are no non-passive listeners:

https://github.com/web-platform-tests/wpt/pull/34623

zcorpan committed:
Test whether events are cancelable with explicit passive/non-passive listeners instead

zcorpan committed:
Rename files to avoid confusion with default passive

zcorpan committed:
Test both passive and non-passive for each event target

zcorpan commented:
@domenic OK I've changed this PR to isolate testing of `cancelable` when there are passive listeners vs when there are non-passive listeners.

domenic commented on dom/events/non-cancelable-when-passive/non-passive-mousewheel-event-listener-on-body.html:
```diff
@@ -0,0 +1,21 @@
+<!DOCTYPE html>
+<title>non-passive mousewheel event listener on body</title>
+<link rel="help" href="https://github.com/WICG/interventions/issues/64">
```
We should ideally have a real spec for this instead of a closed interventions issue...

zcorpan replied:
We can now point to https://w3c.github.io/uievents/#cancelability-of-wheel-events and https://github.com/w3c/uievents/issues/331

domenic commented on dom/events/non-cancelable-when-passive/non-passive-mousewheel-event-listener-on-body.html:
```diff
@@ -0,0 +1,21 @@
+<script src="resources/wait-for.js"></script>
+<script src="resources/scrolling.js"></script>
+<script>
+  onload = () => {
```
I thought if you called promise_test during onload you needed to use explicit_done? I guess I am probably just wrong, the wpt.fyi runs show that the tests are fine, so feel free to resolve... I am just a bit confused.

Is there a particular reason to delay for onload, though? Since everything is a sync script it should be basically equivalent... maybe if you move the `<div>` before, it would be unnecessary.

zcorpan replied:
That sounds familiar. Maybe something changed in testharness.js? I can change the test to avoid onload though.

domenic reviewed (commented):
Generally looks good. However, I do think we should expand the test coverage in two ways:

- We should also test `touchend`, and show that this does *not* apply to `touchend`. Right now it seems an implementation could pass by just applying this to every event.

- We should test synthetic events (with `dispatchEvent()`). Ideally test both the proper event type (e.g. `TouchEvent`) and a non-correct event type (e.g. `Event`). I don't actually know what the behavior will be for these...

There are other cases that might be worth testing, such as:

- `passive: undefined`. Right now we are assuming that the behavior of https://github.com/web-platform-tests/wpt/pull/34623 means `passive: undefined` is the same as `passive: true` for some events, but it feels like this area is messy enough that assuming everything is compositional and can be tested independently is not entirely justified.

- one `passive: true` listener and one `passive: false` listener. (Should behave the same as one `passive: false` listener, IIUC.)

However I don't think these latter are worth testing if they cause a bunch of new files to get generated and need maintenance. If you can merge some files, e.g. test all of `passive: true`, `passive: false`, and `passive: undefined` in a single file, or test all of the events in a single file, or both, then expanding the test matrix might be reasonable. But maybe that is a bad idea for some reason, I am not sure.

zcorpan commented:
This intervention might apply to `touchend`, actually. It's not very clear in the spec what the expected behavior is, but it says that cancelable varies, so maybe it should apply. I filed https://github.com/w3c/touch-events/issues/133 to get some clarity here.

zcorpan commented:
> - We should test synthetic events (with `dispatchEvent()`). Ideally test both the proper event type (e.g. `TouchEvent`) and a non-correct event type (e.g. `Event`). I don't actually know what the behavior will be for these...

I don't think there's anything special going on for synthetic events, but we can test it to verify that that is indeed the case.

zcorpan committed:
Fix help links

zcorpan committed:
Remove onload

zcorpan committed:
Synthetic events are always cancelable by default

zcorpan commented:
I think I've addressed your comments @domenic.

- Testing `touchend` is blocked on https://github.com/w3c/touch-events/issues/133, we can follow up with new tests for this later.
- Test synthetic events: added in 64aca24 
- `passive: undefined` is covered in https://github.com/web-platform-tests/wpt/pull/34623 , I don't think we need to also test it here.
- one `passive: true` listener and one `passive: false` listener, I think is not worth it since it would either make the tests do more things at once or would increase the number of tests for marginal value.

domenic commented on dom/events/non-cancelable-when-passive/synthetic-events-cancelable.html:
```diff
@@ -0,0 +1,34 @@
+    const dispatchEventReturnValue = dispatchEvent(event);
+    assert_false(defaultPrevented, 'defaultPrevented');
+    assert_true(dispatchEventReturnValue, 'dispatchEvent() return value');
+  }, `Synthetic ${eventName} event with interface ${interfaceName} is cancelable`);
```
Isn't this showing the event is *not* cancelable?

zcorpan replied:
Oops
```suggestion
  }, `Synthetic ${eventName} event with interface ${interfaceName} is not cancelable`);
```

domenic reviewed (approved):
LGTM with test name/title nit. However upon reviewing the synthetic-events-cancelable test I realized it was kind of silly. Since you are on the one dispatching the event then it makes sense it will just follow the normal default. And since the normal default is cancelable = false then of course there's no chance for the intervention to apply.

So, feel free to delete that extra test if you want, now that I realize it was not really a tricky case after all.

zcorpan committed:
Apply suggestions from code review

zcorpan commented:
Suggested commit message
```
DOM: Events are non-cancelable when no non-passive listeners

Test that `wheel`, `mousewheel`, `touchstart`, `touchmove` events are not cancelable when there are no non-passive event listeners, and cancelable when there are non-passive event listeners.

Follows https://w3c.github.io/touch-events/#cancelability and https://w3c.github.io/uievents/#cancelability-of-wheel-events

The `mousewheel` event is not yet defined: https://github.com/w3c/uievents/issues/331
```

zcorpan commented:
CI failed to install Safari TP:

> WARNING:install:Did not find any STP URLs

https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=83977&view=logs&jobId=44ec0541-331f-516a-1683-b860fdd2ecd9&j=44ec0541-331f-516a-1683-b860fdd2ecd9&t=90ea950a-2ef3-5f6e-551d-e9a4bef4254f

I tried rebasing but it didn't help. cc @web-platform-tests/admins

gsnedders commented:
> CI failed to install Safari TP:
> 
> > WARNING:install:Did not find any STP URLs
> 
> https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=83977&view=logs&jobId=44ec0541-331f-516a-1683-b860fdd2ecd9&j=44ec0541-331f-516a-1683-b860fdd2ecd9&t=90ea950a-2ef3-5f6e-551d-e9a4bef4254f
> 
> I tried rebasing but it didn't help. cc @web-platform-tests/admins

https://github.com/web-platform-tests/wpt/commit/3fc0804c9c13eafc5356509887b4a1a50c3a2423 should fix that

zcorpan commented:
Thanks @gsnedders!
